### PR TITLE
Dlg string distance

### PR DESCRIPTION
**Pull request by shadrackkibet**
@africanmathsinitiative/developers .This is taking a good shape now and i am still working on it.

### CODE REVIEW

**Comment by shadrackkibet** [comment]
@africanmathsinitiative/developers this is ready for review.

File: instat/dlgStringDistance.vb [hunk 1]
```diff
@@ -32,24 +34,68 @@ Public Class dlgStringDistance
...
+        Dim dctMethod As New Dictionary(Of String, String)
+        dctMethod.Add("Optimal string aligment", Chr(34) & "osa" & Chr(34))
+        dctMethod.Add("Levenshtein distance", Chr(34) & "lv" & Chr(34))
+        dctMethod.Add("Full Damerau-Levenshtein distance", Chr(34) & "dl" & Chr(34))
+        dctMethod.Add("Hamming distance", Chr(34) & "hamming" & Chr(34))
+        dctMethod.Add("Longest common substring distance", Chr(34) & "lcs" & Chr(34))
+        dctMethod.Add("q-gram distance", Chr(34) & "qgram" & Chr(34))
+        dctMethod.Add("cosine distance between q-gram profiles", Chr(34) & "cosine" & Chr(34))
+        dctMethod.Add("Jaccard distance between q-gram profiles", Chr(34) & "jaccard" & Chr(34))
+        dctMethod.Add("Jaro ,or Jaro-Winker distance", Chr(34) & "jw" & Chr(34))
```
Comment (lilyclements) [inline instat/dlgStringDistance.vb]: why is there a comma before the "o" in "or" here?

File: instat/dlgStringDistance.vb [hunk 2]
```diff
@@ -32,24 +34,68 @@ Public Class dlgStringDistance
...
+        dctMethod.Add("Distance based on soundex encoding", Chr(34) & "soundex" & Chr(34))
+        ucrInputComboBoxMethod.SetItems(dctMethod)
 
+        'ucrSave
+        ucrSaveStringDistance.SetPrefix("Dist")
+        ucrSaveStringDistance.SetSaveTypeAsColumn()
+        ucrSaveStringDistance.SetDataFrameSelector(ucrSelectorStringDistance.ucrAvailableDataFrames)
+        ucrSaveStringDistance.SetIsTextBox()
+        ucrSaveStringDistance.SetLabelText("Save Result")
```
Comment (lilyclements) [inline instat/dlgStringDistance.vb]: You need to put a colon after "Save Result" as it leads onto an object which is permanently there (the text box)

**Comment by shadrackkibet** [comment]
@lilyclements i have addressed the changes you suggested.You can make the finicky changes before @dannyparsons reviews this.

File: instat/dlgStringDistance.Designer.vb [hunk 3]
```diff
@@ -22,31 +22,130 @@ Partial Class dlgStringDistance
...
+        '
+        'UcrSelectorStringDistance
+        '
+        Me.UcrSelectorStringDistance.bShowHiddenColumns = False
+        Me.UcrSelectorStringDistance.bUseCurrentFilter = True
+        Me.UcrSelectorStringDistance.Location = New System.Drawing.Point(10, 10)
+        Me.UcrSelectorStringDistance.Margin = New System.Windows.Forms.Padding(0)
+        Me.UcrSelectorStringDistance.Name = "UcrSelectorStringDistance"
+        Me.UcrSelectorStringDistance.Size = New System.Drawing.Size(210, 180)
+        Me.UcrSelectorStringDistance.TabIndex = 0
```
Comment (dannyparsons) [inline instat/dlgStringDistance.Designer.vb]: lower case "u" should be used here.

Comment (lilyclements) [inline instat/dlgStringDistance.Designer.vb]: Damn, how did I miss that!

File: instat/dlgStringDistance.vb [hunk 4]
```diff
@@ -32,24 +33,69 @@ Public Class dlgStringDistance
...
+    Private Sub SetDefaults()
+        clsStringDistFunction = New RFunction
+        ucrInputPatternStringDistance.SetName("")
+        ucrSelectorStringDistance.Reset()
+        ucrSaveStringDistance.Reset()
+        ucrInputComboBoxMethod.Reset()
+        ucrInputPatternStringDistance.Reset()
+        clsStringDistFunction.SetPackageName("stringdist")
+        clsStringDistFunction.SetRCommand("stringdist")
+        clsStringDistFunction.AddParameter("method", Chr(34) & "hamming" & Chr(34))
```
Comment (dannyparsons) [inline instat/dlgStringDistance.vb]: Just wondering why you chose this as the default?

File: instat/dlgStringDistance.Designer.vb (hunk 3 above)
Comment (shadrackkibet) [inline instat/dlgStringDistance.Designer.vb]: I have addressed this minor change.

File: instat/dlgStringDistance.vb (hunk 4 above)
Comment (shadrackkibet) [inline instat/dlgStringDistance.vb]: The default is supposed to be "osa" .Thanks.

**Comment by shadrackkibet** [comment]
@dannyparsons  this is now ready for you.

File: instat/dlgStringDistance.vb [hunk 5]
```diff
@@ -32,24 +33,69 @@ Public Class dlgStringDistance
...
+        dctMethod.Add("Levenshtein Distance", Chr(34) & "lv" & Chr(34))
+        dctMethod.Add("Full Damerau-Levenshtein Distance", Chr(34) & "dl" & Chr(34))
+        dctMethod.Add("Hamming Distance", Chr(34) & "hamming" & Chr(34))
+        dctMethod.Add("Longest Common Substring Distance", Chr(34) & "lcs" & Chr(34))
+        dctMethod.Add("q-Gram Distance", Chr(34) & "qgram" & Chr(34))
+        dctMethod.Add("Cosine Distance between q-Gram Profiles", Chr(34) & "cosine" & Chr(34))
+        dctMethod.Add("Jaccard Distance between q-Gram Profiles", Chr(34) & "jaccard" & Chr(34))
+        dctMethod.Add("Jaro or Jaro-Winker Distance", Chr(34) & "jw" & Chr(34))
+        dctMethod.Add("Distance based on Soundex Encoding", Chr(34) & "soundex" & Chr(34))
+        ucrInputComboBoxMethod.SetItems(dctMethod)
```
Comment (dannyparsons) [inline instat/dlgStringDistance.vb]: osa is also the R default, isn't it? You should set that here as well.

Comment (Lunalo) [inline instat/dlgStringDistance.vb]: true, osa is the default method. Still I am not sure of what this dialog should achieve but still playing around with values

Comment (Lunalo) [inline instat/dlgStringDistance.vb]: @dannyparsons 

I think we should be setting in the SetDefaults?

Comment (dannyparsons) [inline instat/dlgStringDistance.vb]: Set what?

Comment (Lunalo) [inline instat/dlgStringDistance.vb]: we should be setting default value for method in SetDefault

Comment (Lunalo) [inline instat/dlgStringDistance.vb]: but not Initialise

Comment (lilyclements) [inline instat/dlgStringDistance.vb]: If its an R-Default, surely it only needs to be set in InitialiseDialog() only?

Comment (Lunalo) [inline instat/dlgStringDistance.vb]: great Lily. I get your point .

Comment (Lunalo) [inline instat/dlgStringDistance.vb]: so what I was trying to imply is that to allow for proper resetting  of the control was to add the parameter value of osa in SetDefaults and not setting the R default value which is always done in the InitialiseDialog()

I hope this is now clear.

Comment (shadrackkibet) [inline instat/dlgStringDistance.vb]: This discussion has helped me understand somethings which is helpful.I hope all is well for @dannyparsons now.

Comment (dannyparsons) [inline instat/dlgStringDistance.vb]: Great to hear that. Note that now this is the R default we don't need it in the function in set defaults. However, I'm happy to keep it there to confirm it is also the dialog default.

**Pull request merged**